lib/index.js

```diff
@@ -69,70 +79,68 @@ defaults.options = {
+                    return options.onError(request, h, new Boom('No permissions to access this resource', {
+                        statusCode: options.responseCode.onDeny
+                    }));
                 }
 
-                if (!policy || policy === CONFIG_NONE) {
-                    return reply.continue();
+                if (result === RbacCore.UNDETERMINED) {
+                    return options.onError(request, h, new Boom('Could not evaluate access rights to resource', {
+                        statusCode: options.responseCode.onUndetermined
+                    }));
                 }
 
-                // Add context to data retriever's child
-                const wrappedDataRetriever = dataRetriever.createChild(request);
+            } catch (err) {
+                return options.onError(request, h, err);
+            }
+        }
 
-                RbacCore.evaluatePolicy(policy, wrappedDataRetriever, (err, result) => {
+        return h.continue;
+    };
 
-                    if (err) {
-                        return options.onError(request, reply, err);
-                    }
+exports.plugin = {
+    name: Pack.name,
+    version: Pack.version,
+    once: true,
+    multiple: false,
 
-                    if (result === RbacCore.DENY) {
-                        return options.onError(request, reply, Boom.create(options.responseCode.onDeny, 'No permissions to access this resource'));
-                    }
+    register: (server, options) => {
+        Joi.assert(options, schemas.register_options);
 
-                    if (result === RbacCore.UNDETERMINED) {
-                        return options.onError(request, reply, Boom.create(options.responseCode.onUndetermined, 'Could not evaluate access rights to resource'));
-                    }
+        options = Hoek.applyToDefaults(defaults.options, options);
 
-                    reply.continue();
-                });
-            });
-        }
+        // Register default data retrievers
+        const dataRetriever = new DataRetrievalRouter();
+        dataRetriever.register('credentials', require('./dataRetrievers/credentials'));
+        dataRetriever.register('connection', require('./dataRetrievers/connection'));
+        dataRetriever.register('query', require('./dataRetrievers/query-params'));
+        dataRetriever.register(['param', 'params'], require('./dataRetrievers/url-params'));
+        dataRetriever.register('request', require('./dataRetrievers/request'));
 
-        reply.continue();
-    });
+        // Load user defined data retrievers
+        options.dataRetrievers.forEach((dataRetrieverItem) => {
+            dataRetriever.register(dataRetrieverItem.handles, dataRetrieverItem.handler);
+        });
 
-    next();
+        server.ext('onPostAuth', internals.retrievePolicyHandler(dataRetriever, options));
+    }
 };
 
-exports.register.attributes = {
+
+exports.plugin.register.attributes = {
     pkg: require('../package.json')
```

Comment:
Can use the const Pack created above
```suggestion
    pkg: Pack
```